### PR TITLE
fix(rust, python): Remove extra `\n` when reading file-like object wi…

### LINKED ISSUE #12266: Extra null while reading data frame from zip archive
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

The code in test (in a package called investigate), file in question is unzip_polars.py

```python
import zipfile
from io import BytesIO
from typing import Dict, Any

import polars as pl


def save_it_somewhere(df: pl.DataFrame):
    pass


def read_data(data: Dict[str, Any]):
    # Get data - Downloaded as zip file
    zip_data = BytesIO(data["data"])

    # Use ZipFile to extract the contents of the zip file
    with zipfile.ZipFile(zip_data, "r") as zip_file:
        csv_filename = zip_file.namelist()[0]

        # Read the CSV file into a Polars DataFrame
        with zip_file.open(csv_filename) as csv_file:
            df = pl.read_csv(csv_file, infer_schema_length=2000)

    save_it_somewhere(df)
```

The test case file in question is test_unzip_polars.py

```python
import os
import zipfile
from pathlib import Path
from unittest.mock import patch, MagicMock

import polars as pl
from polars.testing import assert_frame_equal

from investigate.unzip_polars import read_data

MY_DATA = pl.DataFrame({"a": [1, 2, 3]})
# MY_DATA = pl.DataFrame({"a": [1, 2, 3], "b": ["x", "y", "z"]})


def zip_a_pandas(tmp_path: Path) -> bytes:
    dir_with_zipfile = tmp_path / "sub"
    dir_with_zipfile.mkdir()
    file_path: Path = dir_with_zipfile / "hello.csv"
    zip_path = tmp_path / "hello.zip"
    MY_DATA.write_csv(file_path)

    # Create object of ZipFile
    with zipfile.ZipFile(zip_path, "w") as zip_object:
        # Traverse all files in directory
        for folder_name, sub_folders, file_names in os.walk(dir_with_zipfile):
            for filename in file_names:
                # Create filepath of files in directory
                some_path = os.path.join(folder_name, filename)
                # Add files to zip file
                zip_object.write(some_path, os.path.basename(some_path))

    with open(zip_path, mode="rb") as zip_file:
        file_bytes = zip_file.read()

    Path(file_path).unlink()
    Path(zip_path).unlink()
    dir_with_zipfile.rmdir()

    return file_bytes


@patch("investigate.unzip_polars.save_it_somewhere")
def test_read_write(saver_mock: MagicMock, tmp_path):
    read_data({"data": zip_a_pandas(tmp_path)})

    assert saver_mock.call_count == 1

    arg_list = saver_mock.call_args_list

    assert_frame_equal(MY_DATA, arg_list[0][0][0])

    tmp_path.rmdir()

```

### Log output

```shell
_________________________________________________________________________________________________________________ test_read_write __________________________________________________________________________________________________________________

saver_mock = <MagicMock name='save_it_somewhere' id='4575202192'>, tmp_path = PosixPath('/private/var/folders/70/tpb1l8jd45s8zj742ljrmbs80000gp/T/pytest-of-vassilisanagnostopoulos/pytest-19/test_read_write0')

    @patch("investigate.unzip_polars.save_it_somewhere")
    def test_read_write(saver_mock: MagicMock, tmp_path):
        read_data({"data": zip_a_pandas(tmp_path)})
    
        assert saver_mock.call_count == 1
    
        arg_list = saver_mock.call_args_list
    
>       assert_frame_equal(MY_DATA, arg_list[0][0][0])

test_unzip_polars.py:50: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

left = shape: (3, 1)
┌─────┐
│ a   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
│ 2   │
│ 3   │
└─────┘, right = shape: (4, 1)
┌──────┐
│ a    │
│ ---  │
│ i64  │
╞══════╡
│ 1    │
│ 2    │
│ 3    │
│ null │
└──────┘

    def assert_frame_equal(
        left: DataFrame | LazyFrame,
        right: DataFrame | LazyFrame,
        *,
        check_row_order: bool = True,
        check_column_order: bool = True,
        check_dtype: bool = True,
        check_exact: bool = False,
        rtol: float = 1e-5,
        atol: float = 1e-8,
        categorical_as_str: bool = False,
        nans_compare_equal: bool | None = None,
    ) -> None:
        """
        Assert that the left and right frame are equal.
    
        Raises a detailed ``AssertionError`` if the frames differ.
        This function is intended for use in unit tests.
    
        Parameters
        ----------
        left
            The first DataFrame or LazyFrame to compare.
        right
            The second DataFrame or LazyFrame to compare.
        check_row_order
            Require row order to match.
    
            .. note::
                Setting this to ``False`` requires sorting the data, which will fail on
                frames that contain unsortable columns.
        check_column_order
            Require column order to match.
        check_dtype
            Require data types to match.
        check_exact
            Require data values to match exactly. If set to ``False``, values are considered
            equal when within tolerance of each other (see ``rtol`` and ``atol``).
            Logical types like dates are always checked exactly.
        rtol
            Relative tolerance for inexact checking. Fraction of values in ``right``.
        atol
            Absolute tolerance for inexact checking.
        categorical_as_str
            Cast categorical columns to string before comparing. Enabling this helps
            compare columns that do not share the same string cache.
        nans_compare_equal
            Consider NaN values to be equal.
    
            .. deprecated: 0.19.12
                This parameter will be removed. Default behaviour will remain as though it
                were set to ``True``.
    
        See Also
        --------
        assert_series_equal
        assert_frame_not_equal
    
        Notes
        -----
        When using pytest, it may be worthwhile to shorten Python traceback printing
        by passing ``--tb=short``. The default mode tends to be unhelpfully verbose.
        More information in the
        `pytest docs <https://docs.pytest.org/en/latest/how-to/output.html#modifying-python-traceback-printing>`_.
    
        Examples
        --------
        >>> from polars.testing import assert_frame_equal
        >>> df1 = pl.DataFrame({"a": [1, 2, 3]})
        >>> df2 = pl.DataFrame({"a": [1, 5, 3]})
        >>> assert_frame_equal(df1, df2)  # doctest: +SKIP
        Traceback (most recent call last):
        ...
        AssertionError: Series are different (value mismatch)
        [left]:  [1, 2, 3]
        [right]: [1, 5, 3]
    
        The above exception was the direct cause of the following exception:
    
        Traceback (most recent call last):
        ...
        AssertionError: values for column 'a' are different
    
        """
        if nans_compare_equal is not None:
            issue_deprecation_warning(
                "The `nans_compare_equal` parameter for `assert_frame_equal` is deprecated."
                " Default behaviour will remain as though it were set to `True`.",
                version="0.19.12",
            )
        else:
            nans_compare_equal = True
    
        lazy = _assert_correct_input_type(left, right)
        objects = "LazyFrames" if lazy else "DataFrames"
    
        _assert_frame_schema_equal(
            left,
            right,
            check_column_order=check_column_order,
            check_dtype=check_dtype,
            objects=objects,
        )
    
        if lazy:
            left, right = left.collect(), right.collect()  # type: ignore[union-attr]
        left, right = cast(DataFrame, left), cast(DataFrame, right)
    
        if left.height != right.height:
>           raise_assertion_error(
                objects, "number of rows does not match", left.height, right.height
            )
E           AssertionError: DataFrames are different (number of rows does not match)
E           [left]:  3
E           [right]: 4

../.pyenv/versions/3.11.6/lib/python3.11/site-packages/polars/testing/asserts/frame.py:122: AssertionError
--------------------------------------------------------------------------------------------------------------- Captured stderr call ---------------------------------------------------------------------------------------------------------------
Polars found a filename. Ensure you pass a path to the file instead of a python file object when possible for best performance.
============================================================================================================= short test summary info ==============================================================================================================
FAILED test_unzip_polars.py::test_read_write - AssertionError: DataFrames are different (number of rows does not match)
================================================================================================================ 1 failed in 0.36s =================================================================================================================
➜  investigate
```

### Issue description

If I comment first dataframe and uncomment the second one test works.

First case while reading adds a null row.

### Expected behavior

I would expect both cases to work.

### Installed versions

<details>

```
>>> pl.show_versions()
--------Version info---------
Polars:              0.19.12
Index type:          UInt32
Platform:            macOS-13.5.2-x86_64-i386-64bit
Python:              3.11.6 (main, Oct 23 2023, 17:15:48) [Clang 15.0.0 (clang-1500.0.40.1)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
gevent:              <not installed>
matplotlib:          3.8.1
numpy:               1.26.1
openpyxl:            <not installed>
pandas:              2.1.2
pyarrow:             <not installed>
pydantic:            1.10.13
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          2.0.22
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>

```

</details>

### LINKED ISSUE #12404: Raise ComputeError on reading non-LF eol csv as file-like object
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
import io

csv = b"a;1;2;"
# Using BufferedReader is to reproduce in memory. FileIO can also cause this.
buf = io.BufferedReader(io.BytesIO(csv)) # type: ignore[arg-type]
df = pl.read_csv(buf, eol_char=";")
```

### Log output

```shell
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "~\AppData\Roaming\Python\Python311\site-packages\polars\io\csv\functions.py", line 366, in read_csv
    df = pl.DataFrame._read_csv(
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "~\AppData\Roaming\Python\Python311\site-packages\polars\dataframe\frame.py", line 775, in _read_csv
    self._df = PyDataFrame.read_csv(
               ^^^^^^^^^^^^^^^^^^^^^
polars.exceptions.ComputeError: Could not parse `
` as dtype `i64` at column 'a' (column number 1).
The current offset in the file is 6 bytes.

You might want to try:
- increasing `infer_schema_length` (e.g. `infer_schema_length=10000`),
- specifying correct dtype with the `dtypes` argument
- setting `ignore_errors` to `True`,
- adding `
` to the `null_values` list.
```

### Issue description

* read as bytes -> OK
* read as BytesIO -> OK
* read as BinaryIO (BufferedReader, FileIO, etc.) -> Error

Related: #12266
The cause of 12266 is that LF is added even if the file is ends with LF.
And the cause of this ticket is that LF is added even if the eol char is not LF.

I created PR to fix the extra-null issue on #12333, but the PR did not cover this compute error issue.
So I have updated the PR to cover this issue.

### Expected behavior

Success to read.

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.13
Index type:          UInt32
Platform:            Windows-10-10.0.22631-SP0
Python:              3.11.6 (tags/v3.11.6:8b6ee5b, Oct  2 2023, 14:57:12) [MSC v.1935 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         2.2.0
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              2022.11.0
gevent:              <not installed>
matplotlib:          3.8.0
numpy:               1.26.1
openpyxl:            <not installed>
pandas:              2.1.3
pyarrow:             14.0.1
pydantic:            <not installed>
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          3.1.2
```

</details>

### PR DESCRIPTION
**Pull request by hirohira9119**
…th `read_csv`

To fix #12266 
To fix #12404

In the code before the fix, the conditional statement was always evaluated to true. So, for a file-like object, even if the file ended with a '\n', an extra '\n' was always added (the result was like `b"a\n1\n\n"` or `b"a,b\n1,1\n\n"`).
Then, in case of single column, the repeated newlines were interpreted as if there was missing value.
This was the extra null.

On the other hands, in the case of multiple columns, they were not treated as if there were missing values. 
So this issue only appeared when dealing with single-column files as file-like object.

[edit]
I found another issue caused by this. Because the condition ignores the `eol_char`, `Compute Error` is raised when `eol_char` is set.
The previous PR does not cover the case of non-LF eol-char.
I think we can remove the part of adding eol-char because it works without one and also the other branches works without the equivalent part.
And the remove solves the both issues. So I've updated the PR.

### CODE REVIEW

File: py-polars/tests/unit/io/files/single_column.csv [hunk 1]
```diff
@@ -0,0 +1,4 @@
+a
```
Comment (ritchie46) [inline py-polars/tests/unit/io/files/single_column.csv:1]: Can you make this file in the test?

You can put that string in a file buffer and read that in the test.

**Review by ritchie46 — CHANGES_REQUESTED** [review]

File: py-polars/tests/unit/io/files/single_column.csv (hunk 1 above)
Comment (hirohira9119) [reply py-polars/tests/unit/io/files/single_column.csv]: Yes, I have made the commit for the change.
I also noticed another thing that needs testing, when the file ends without LF. So, I've added that to the update.

File: py-polars/tests/unit/io/test_csv.py [hunk 2]
```diff
@@ -1639,3 +1639,57 @@ def write(self, data: bytes) -> int:
     df.write_csv(buf)  # type: ignore[call-overload]
 
     assert b"".join(buf.writes) == b"a,b\n10,x\n20,y\n30,z\n"
+
+
+@pytest.mark.write_disk()
+def test_read_filelike_object(tmp_path: Path) -> None:
```
Comment (ritchie46) [inline py-polars/tests/unit/io/test_csv.py:1645]: Though better. I think we can do this test completely in memory by using `io.Buffer`

Comment (hirohira9119) [reply py-polars/tests/unit/io/test_csv.py]: Sorry for misunderstood. I confirmed that this issue is also caused with `io.BufferedReader` and this PR fixed it.
So I have committed the update.

**Comment by hirohira9119** [comment]
I have updated the PR to cover the another issue that I found.

**Review by ritchie46 — APPROVED** [review]

**Comment by ritchie46** [comment]
Thank you @hirohira9119

**Pull request merged**